contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,527 @@
+    VRFConsumerV2Interface v;
+    bytes memory resp = abi.encodeWithSelector(v.fulfillRandomWords.selector, requestId, randomWords);
+    uint256 gasPreCallback = gasleft();
+    if (gasPreCallback < fp.callbackGasLimit) {
+      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
+    }
+    (bool success,) = fp.sender.call(resp);
+    // Avoid unused-local-variable warning. (success is only present to prevent
+    // a warning that the return value of consumerContract.call is unused.)
+    (success);
+
+    emit RandomWordsFulfilled(requestId, randomWords, success);
+    // We want to charge users exactly for how much gas they use in their callback.
+    // The gasAfterPaymentCalculation is meant to cover these additional operations where we
+    // decrement the subscription balance and increment the oracles withdrawable balance.
+    uint96 payment = calculatePaymentAmount(startGas, s_config.gasAfterPaymentCalculation, tx.gasprice);
+    if (s_subscriptions[fp.subId].balance < payment) {
+      revert InsufficientBalance();
+    }
+    s_subscriptions[fp.subId].balance -= payment;
+    s_withdrawableTokens[s_serviceAgreements[keyHash]] += payment;
+  }
+
+  // Get the amount of gas used for fulfillment
+  function calculatePaymentAmount(
+      uint256 startGas,
+      uint256 gasAfterPaymentCalculation,
+      uint256 weiPerUnitGas
+  )
+    private
+    view
+    returns (
+      uint96
+    )
+  {
+    uint256 weiPerUnitLink;
+    weiPerUnitLink = getFeedData();
+    if (weiPerUnitLink < 0) {
+      revert InvalidFeedResponse(weiPerUnitLink);
+    }
+    // (1e18 jules/link) (wei/gas * gas) / (wei/link) = jules
+    return uint96(1e18*weiPerUnitGas*(gasAfterPaymentCalculation + startGas - gasleft()) / weiPerUnitLink);
+  }
+
+  function getRandomnessFromProof(
+    bytes memory proof
+  )
+    public 
+    view 
+    returns (
+      bytes32 currentKeyHash,
+      uint256 requestId, 
+      uint256 randomness, 
+      FulfillmentParams memory fp
+    ) 
+  {
+    // blockNum follows proof, which follows length word (only direct-number
+    // constants are allowed in assembly, so have to compute this in code)
+    uint256 blockNumOffset = 0x20 + PROOF_LENGTH;
+    // Note that proof.length skips the initial length word.
+    // We expected the total length to be proof + 5 words (blocknum, subId, callbackLimit, nw, sender)
+    if (proof.length != PROOF_LENGTH + 0x20*5) {
+      revert InvalidProofLength(proof.length, PROOF_LENGTH + 0x20*5);
+    }
+    uint256[2] memory publicKey;
+    uint256 preSeed;
+    uint256 blockNum;
+    address sender;
+    assembly { // solhint-disable-line no-inline-assembly
+      publicKey := add(proof, PUBLIC_KEY_OFFSET)
+      preSeed := mload(add(proof, PRESEED_OFFSET))
+      blockNum := mload(add(proof, blockNumOffset))
+      // We use a struct to limit local variables to avoid stack depth errors.
+      mstore(fp, mload(add(add(proof, blockNumOffset), 0x20)))
```

Comment:
How many words are there in a `FulfillmentParams`? I count two ((3*64+160)/256=1.4), but this seems to be copying across 3. It'd be helpful to have comments about which portions of `FulfillmentParams` each of these `mstore`s is responsible for.

Reply:
Yep will add a comment. In memory structs aren't packed like that (only storage structs are) so subId, callbackGasLimit and numWords are all in their own words